### ISSUE TITLE
TimeControl Time Skip Bug

### ISSUE THREAD

**Issue by petethered** [kitten-science/kitten-scientists#66]
https://github.com/kitten-science/kitten-scientists/blob/2ee3168d4ddc8af36b1aea9febb9cef604a443cc/packages/userscript/source/TimeControlManager.ts#L377

If **settings.timeSkip.max** = -1 (infinite in the interface), **canSkip** is ALWAYS -1 thanks to the referenced line above.

The next conditional is:

```
   if (canSkip < remainingYearsCurrentCycle) {
      willSkip = canSkip;
```

Since **canSkip** is -1, **willSkip** will get set to -1 also, effectively DISABLING time skipping.

Something like:

```
// if infinite & not overheating
if (this.settings.timeSkip.max == -1 && Math.floor((heatMax - heatNow) / factor) > 0) { 
    // canSkip up to the heatMax or the amount of time crystals
    canSkip = Math.min(Math.floor((heatMax - heatNow) / factor), Math.floor(timeCrystal.value)); 
}
```

Following line 377 will fix it.

**Comment by petethered** [comment]
As a side note:

https://github.com/kitten-science/kitten-scientists/blob/2ee3168d4ddc8af36b1aea9febb9cef604a443cc/packages/userscript/source/TimeControlManager.ts#L340

If you are in a temporal paradox, at least in the interface of kittens game, the day is negative.

**Comment by oliversalzburg** [comment]
Nice. Thanks a lot for already having debugged this! I will try to include the fix ASAP.

Just FYI, I feel like the problem is related to a common pattern in infinity handling. Code like this is in several places of the codebase:
```ts
const maxPastures =
  -1 === this.settings.buildings.unicornPasture.max
    ? Number.POSITIVE_INFINITY
    : this.settings.buildings.unicornPasture.max;
```

It is likely that this piece of code needs to be generalized and applied at all relevant locations.

**Comment by petethered** [comment]
Thanks!

I was trying to figure out why it wasn't skipping and dug through the code, so figured I'd spare you the work of finding it yourself.

Since you will be in there, there's 2 more things you might want to "tweak":

I didn't dig through the code to verify, but I have a feeling it also won't do a "resource control check" before initiating the skip.  So if you have a resource limit set on time crystals, it might go past that limit.

<img width="483" alt="image" src="https://user-images.githubusercontent.com/1239247/207677711-2147c1a6-c33e-419f-a0ba-7b02fe33a60b.png">

The trigger message for time control doesn't make sense.  It's using this value to compare against the number of time crystals you have on line 347

---

Love the scientists, thanks for your effort!

**Comment by oliversalzburg** [comment]
Nice! I'll try to get those fixed right now.

BTW, I feel like the first mentioned issue relates to this feature request:
- #18 

If max heat is configured to infinite, max heat should be ignored and that is also the fix for this problem. Or am I missing something?

**Comment by oliversalzburg** [comment]
Regarding the resource behavior: I assume what you're observing is due to the fact that the time skip automation runs last in the KS processing loop. So the trigger value might not have been reached at the end of a tick, you gain additional crystals at the start of the next tick and then it triggers the time skip.

Let me know if you think this could explain what you're seeing. Otherwise I might need a save where I can reproduce the behavior.

Regarding the trigger: Not only does the message not make any sense when setting the trigger value, it also seems to behave incorrectly entirely. I created a dedicated issue for this:
- #67

**Comment by petethered** [comment]
[Kittens Game - Run 16 - Year 21 - Spring, day 15.txt](https://github.com/kitten-science/kitten-scientists/files/10231689/Kittens.Game.-.Run.16.-.Year.21.-.Spring.day.15.txt)

Check resource control, make sure time crystal is set to 1740

<img width="332" alt="image" src="https://user-images.githubusercontent.com/1239247/207715283-f2a439af-ba52-428b-ae4f-6f6badb878b4.png">

Then turn on time control

<img width="383" alt="image" src="https://user-images.githubusercontent.com/1239247/207715317-efc96c14-e5c9-44e0-8ef2-af7db7d5dae5.png">

And it will start skipping years , taking time crystal below the limit.

<img width="364" alt="image" src="https://user-images.githubusercontent.com/1239247/207715415-0fd341f7-08e3-4559-aed9-0caeeb7511ed.png">

Basically, after you set **canSkip**, you need to have it check if TC is under resource control and adjust canSkip to TC - limit

**Comment by petethered** [comment]
I'm loading kitten via bookmarklet:

javascript:(function(){var d=document,s=d.createElement('script');s.src='https://github.com/kitten-science/kitten-scientists/releases/download/v2.0.0-alpha.10/kitten-scientists-2.0.0-alpha.10.user.js';d.body.appendChild(s);})();

**Comment by petethered** [comment]
#18

Interestingly, his diff patch does NOT solve the canSkip bug above.  It just sets **canSkip** to the timeSkip.max, which means that he must have set a value in the ui for timeSkip.max for him to be skipping at all.  If I were to guess, I'd say he has it set to 5 or lower as he doesn't patch the "willSkip" loop as it recalculates heat.

The requested "ignore heat" should probably be a toggle-able option as his specific use case (fast paragon 600 runs) can safely ignore heat, but the general case where you are trying to maximize your ROI on crystals means you want to pay attention to the heat (as is in the code already).

**Comment by oliversalzburg** [comment]
> Basically, after you set canSkip, you need to have it check if TC is under resource control and adjust canSkip to TC - limit

There's actually an earlier check when evaluating the trigger, which doesn't take resource control into account. This will be fixed with the other time control issues.

**Comment by petethered** [comment]
You probably want to change:

`-1 === this.settings.timeSkip.max ? Number.POSITIVE_INFINITY : this.settings.timeSkip.max;`

to 

`-1 === this.settings.timeSkip.max ? timeCrystalsAvailable : this.settings.timeSkip.max;`

In your pull, otherwise you only partially fix the "resource control" issue.

**Comment by petethered** [comment]
Here's the hypothetical/test case:

```const timeCrystalsAvailable = this._workshopManager.getValueAvailable("timeCrystal")```

If **timeCrystalsAvailable** == 1 post calculation , but **this.settings.timeSkip.max** == -1, minHeat will get to **Number.POSITIVE_INFINITY** allowing it to go past the available crystals.

You could alternatively change 

```let canSkip = Math.min(Math.floor((heatMax - heatNow) / factor), maxHeat);```

to

```let canSkip = Math.min(Math.floor((heatMax - heatNow) / factor), maxHeat, Math.floor(timeCrystalsAvailable));```

If you want to preserve the structure of the infinite code and let the min function solve it for you

(I keep wrapping timeCrystalsAvailable in a Math.floor given that I don't know if an integer value comes out of the workshopcode)

**Comment by oliversalzburg** [comment]
Sorry, I somehow missed your last comments before I merged the PR.

I'm not sure I understand your analysis correctly. I will likely have to debug this with your save to get a better understanding of the situation.

**Comment by oliversalzburg** [comment]
I tried your save and I see myself ending up with 1736.651 time crystals pretty much instantly (in case you're not aware, your KS settings are already in the save). 

If I understand you correctly, you want that, when time skipping, you will never fall below the configured stock value for time crystals. Thus, KS should only time skip when you have enough time crystals that you will still be above your configured stock value after the time skip.

Is that correct?

**Comment by petethered** [comment]
Re: Math.floor()

I just put it in my code samples there because I haven't looked through the code to see when an int gets returned or not.  I'm not sure when the code requires an int and when a float can slip through

In reviewing getValueAvailable , it returns an int as the return is either the result of Math.max or Math.min (which return int)

So it's safe to ignore those previously used ones from me.

Re: Resource control

First, an assumption/ux decision:  If a resource is being limited by resource control, NOTHING else in KS should be allowed to go below that number (if possible as I'm pretty sure you can't limit faith).  

That makes logical sense to me, what I assumed would happen before looking at the source as "just a user", and seems to be pretty consistently applied across other functions.  For example, if titanium is limited in resource control, bonfire , space, and workshop (for upgrades) obeys that limit.

If this is correct, then Time Control needs to follow this behavior as well, not allowing time crystals drop below the allowed amount.

https://github.com/kitten-science/kitten-scientists/blob/83e93aae807ba5fbb3e8dddadb23bf67925c8444/packages/userscript/source/TimeControlManager.ts#L382

That's currently 

```      -1 === this.settings.timeSkip.max ? Number.POSITIVE_INFINITY : this.settings.timeSkip.max;```

So if timeSkip.max is set to infinity it will consume as many time crystals as it wants.

If it timeSkip.max is set to a value it will consume that much (assuming you have the "heat" for it).

Both of these can cause the time crystals to go below the resource controlled limit.  

The line needs to change to:

```-1 === this.settings.timeSkip.max ? timeCrystalsAvailable : Math.min(this.settings.timeSkip.max, timeCrystalsAvailable);```

If infinite, it will use up to the amount of crystals available after resource control.

If restricted, it will use the min of the max and timeCrystalsAvailable.

**Comment by petethered** [comment]
Here's 2 hypotheticals:

The player has 100 time crystals

It's RC'd to 99

timeSkip.max = -1

They have 0 heat and no upgrades, so their limit is 300 and the factor = 10 

https://github.com/kitten-science/kitten-scientists/blob/83e93aae807ba5fbb3e8dddadb23bf67925c8444/packages/userscript/source/TimeControlManager.ts#L350

sets timeCrystalsAvailable = 1 

```    const maxHeat =
      -1 === this.settings.timeSkip.max ? Number.POSITIVE_INFINITY : this.settings.timeSkip.max;```
      
maxHeat evaluates as      Number.POSITIVE_INFINITY

```let canSkip = Math.min(Math.floor((heatMax - heatNow) / factor), maxHeat); ```

canSkip is now 30

And so the code burns 30 crystals, going below the RC limit and their time crystals are now 70.

Next tick abourts as timeCrystalsAvailable = -30

-----

The player has 100 time crystals

It's RC'd to 99

timeSkip.max = 5  (since they want to skip one cycle per trigger)

They have 0 heat and no upgrades, so their limit is 300 and the factor = 10 

https://github.com/kitten-science/kitten-scientists/blob/83e93aae807ba5fbb3e8dddadb23bf67925c8444/packages/userscript/source/TimeControlManager.ts#L350

Again, timeCrystalsAvailable = 1 

```    const maxHeat =
      -1 === this.settings.timeSkip.max ? Number.POSITIVE_INFINITY : this.settings.timeSkip.max;```
      
Now maxHeat = 5 as that's the max

```let canSkip = Math.min(Math.floor((heatMax - heatNow) / factor), maxHeat); ```

canSkip = 5

it skips

time crystals = 95 now

------------

In both of those cases, my proposed alteration

```-1 === this.settings.timeSkip.max ? timeCrystalsAvailable : Math.min(this.settings.timeSkip.max, timeCrystalsAvailable);```

results in maxHeat = 1

A single skip happens and the player has 99 time crystals afterwords, their limit value

**Comment by petethered** [comment]
Again, thanks for your attention and efforts!

**Comment by oliversalzburg** [comment]
> If this is correct, then Time Control needs to follow this behavior as well, not allowing time crystals drop below the allowed amount.
Agreed.

I prefer the suggested fix on the `canSkip` expression though. I'm currently testing it.
